src/Interactive/Host/Interactive/Core/InteractiveHost.Service.cs

```diff
@@ -367,44 +358,42 @@ private static string GenerateUniqueChannelLocalName()
             /// Reads given initialization file (.rsp) and loads and executes all assembly references and files, respectively specified in it.
             /// Execution is performed on the UI thread.
             /// </summary>
-            [OneWay]
-            public void InitializeContext(RemoteAsyncOperation<RemoteExecutionResult> operation, string initializationFile, bool isRestarting)
+            public void InitializeContext(string initializationFile, bool isRestarting)
             {
-                Debug.Assert(operation != null);
 
+                var completionSource = new TaskCompletionSource<RemoteExecutionResult>();
                 lock (_lastTaskGuard)
                 {
-                    _lastTask = InitializeContextAsync(_lastTask, operation, initializationFile, isRestarting);
+                    _lastTask = InitializeContextAsync(_lastTask, completionSource, initializationFile, isRestarting);
                 }
             }
 
             /// <summary>
             /// Adds an assembly reference to the current session.
             /// </summary>
-            [OneWay]
-            public void AddReference(RemoteAsyncOperation<bool> operation, string reference)
+            public void AddReference(string reference)
             {
-                Debug.Assert(operation != null);
                 Debug.Assert(reference != null);
 
                 lock (_lastTaskGuard)
                 {
-                    _lastTask = AddReferenceAsync(_lastTask, operation, reference);
+                    // (miziga) change operation to completionSource? Use JsonRpc in InteractiveHost instead?
+                    _lastTask = AddReferenceAsync(_lastTask, reference);
                 }
             }
 
-            private async Task<EvaluationState> AddReferenceAsync(Task<EvaluationState> lastTask, RemoteAsyncOperation<bool> operation, string reference)
+            private async Task<EvaluationState> AddReferenceAsync(Task<EvaluationState> lastTask, string reference)
```

Comment:
>lastTask [](start = 88, length = 8)

You can use `TaskCompletionSource<bool>` here as you use it in the other methods.

Reply:
for the finally block of this method, do we then set the result of completionSource to the success boolean since the CompleteExecution method takes a TaskCompletionSource<RemoteExecutionResult>?

Reply:
Yup

---
In reply to: [431370974](https://github.com/dotnet/roslyn/pull/44434#discussion_r431370974) [](ancestors = 431370974)